toolchain/driver/lld_subcommand.cpp

```diff
@@ -0,0 +1,109 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/driver/lld_subcommand.h"
+
+#include "llvm/TargetParser/Host.h"
+#include "llvm/TargetParser/Triple.h"
+#include "toolchain/driver/lld_runner.h"
+
+namespace Carbon {
+
+auto LLDOptions::Build(CommandLine::CommandBuilder& b) -> void {
+  // We want to select a default platform based on the default target. Since
+  // that requires some dynamic inspection of the target, do that here.
+  std::string default_target = llvm::sys::getDefaultTargetTriple();
+  llvm::Triple default_triple(default_target);
+  switch (default_triple.getObjectFormat()) {
+    case llvm::Triple::MachO:
+      platform = Platform::Darwin;
+      break;
+
+      // We default to the GNU or Unix platform as ELF is a plausible default
+      // and LLD doesn't support any generic invocations.
+    default:
+    case llvm::Triple::ELF:
+      platform = Platform::Gnu;
+      break;
+  }
+
+  b.AddOneOfOption(
+      {
+          .name = "platform",
+          .help = R"""(
+Platform linking style to use. The default is selected to match the default
+target's platform.
+)""",
+      },
+      [&](auto& arg_b) {
+        arg_b.SetOneOf(
+            {
+                arg_b.OneOfValue("gnu", Platform::Gnu),
+                // Some of LLD documentation uses "Unix" instead of "GNU", so
+                // include an alias here.
+                arg_b.OneOfValue("unix", Platform::Gnu),
+                arg_b.OneOfValue("darwin", Platform::Darwin),
+            },
+            &platform);
+      });
+  b.AddStringPositionalArg(
+      {
+          .name = "ARG",
+          .help = R"""(
+Arguments passed to LLD.
+)""",
+      },
+      [&](auto& arg_b) { arg_b.Append(&args); });
+}
+
+static constexpr CommandLine::CommandInfo SubcommandInfo = {
+    .name = "lld",
+    .help = R"""(
+Runs LLD with the provided arguments.
+
+Note that a specific LLD platform must be selected, and it is actually that
+particular platform's LLD-driver that is run with the arguments. There is no
+generic LLD command line.
+
+For a given platform, this is equivalent to running that platform's LLD alias
+directly, and provides the full command line interface.
+
+Use `carbon lld --platform=gnu -- ARGS` to separate the `ARGS` forwarded to LLD
+from the flags passed to the Carbon subcommand.
+
+Note that typically it is better to use a higher level command to link code,
+such as invoking `carbon link` with the relevant flags. However, this subcommand
```

Comment:
Stepping back a moment, comparing with `carbon clang`, my understanding is that we'll recommend `carbon clang` but `clang` will also work as a thin wrapper to that. This feels a bit more in the other direction, designed to be a thin wrapper around `lld`; is that going to change? Why did you choose to add subcommands instead of something like calling `lld_main` directly from `main`?

(maybe helpful to put notes in the PR description for notes, since I expect other things like `cc1` would follow the same pattern?)

Reply:
Yes, separate talk about long-term helped